source/change-streams/change-streams.rst

```diff
@@ -414,25 +462,96 @@ Once a ``ChangeStream`` has encountered a resumable error, it MUST attempt to re
+- If the ``ChangeStream`` has no cached ``documentResumeToken``, and ``resumeAfter`` is not specified, and the max wire version is >= ``7``:
+
+  - The driver MUST execute the known aggregation command.
+  - If the ``ChangeStream`` has a saved ``postBatchResumeToken``:
+
+    - The driver MUST specify ``resumeAfter`` with the cached ``postBatchResumeToken``.
+  - Else:
 
-    - The driver MUST execute the known aggregation command.
     - The driver MUST specify the ``startAtOperationTime`` key set to the ``startAtOperationTime`` provided by the user or saved from the original aggregation.
     - The driver MUST NOT set a ``resumeAfter`` key.
-    - In this case, the ``ChangeStream`` will return all changes that occurred after the specified ``startAtOperationTime``.
 - Else:
 
-    - The driver MUST execute the known aggregation command.
-    - The driver MUST specify a ``resumeAfter`` with the last known ``resumeToken``.
-    - The driver MUST NOT set a ``startAtOperationTime``.
-    - If a ``startAtOperationTime`` key was part of the original aggregation command, the driver MUST remove it.
-    - In this case, the ``ChangeStream`` will return notifications starting with the oplog entry immediately *after* the provided token.
+  - The driver MUST execute the known aggregation command.
+  - If the ``ChangeStream`` has returned all documents in the most recent batch (or the most recent batch was empty) and has a cached ``postBatchResumeToken``:
+
+    - The driver MUST specify a ``resumeAfter`` with the cached ``postBatchResumeToken``.
+  - Else:
+
+    - The driver MUST specify a ``resumeAfter`` with the cached ``documentResumeToken``.
+  - The driver MUST NOT set a ``startAtOperationTime``.
+  - If a ``startAtOperationTime`` key was part of the original aggregation command, the driver MUST remove it.
+  - In this case, the ``ChangeStream`` will return notifications starting with the oplog entry immediately *after* the provided token.
 
 If the server supports sessions, the resume attempt MUST use the same session as the previous attempt's command.
 
 A driver SHOULD attempt to kill the cursor on the server on which the cursor is opened during the resume process, and MUST NOT attempt to kill the cursor on any other server.
 
 
+Exposing All Resume Tokens
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:since: 4.2
+
+Users can retrieve the ``documentResumeToken`` by inspecting the _id on each ``ChangeDocument``. But since MongoDB 4.2, aggregate and getMore responses also include a ``postBatchResumeToken``. Drivers use one or the other when automatically resuming, as described in `Resume Process`_.
+
+Drivers MUST expose a mechanism to retrieve the same resume token that would be used to automatically resume. It MUST be possible to use this mechanism after iterating every document. It MUST be possible for users to use this mechanism periodically even when no documents are getting returned. Drivers have two options to implement this.
```

Comment:
Suggest adding "(i.e. `getMore` has returned empty batches)" after "are getting returned", unless you think it's clear as-is.

Reply:
Agreed. Done.